File: typescript/src/shared/fmt/num.ts
import { displayDecimalsBy, isFormattedZero } from "./constants";

export function displayFormattedString(input: string): string {
  const [rawInteger, rawFraction] = input.split(".");
  const negative = rawInteger.startsWith("-");
  const integer = negative ? rawInteger.slice(1) : rawInteger;
  const withCommas = integer.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  const fraction = rawFraction;
  const prefix = negative ? "-" : "";

  if (fraction === undefined) {
    return `${prefix}${withCommas}`;
  }

  return `${prefix}${withCommas}.${fraction}`;
}

export function displayWithDecimals(value: number, decimals: number): string {
  return displayFormattedString(value.toFixed(decimals));
}

function displayDecimals(absValue: number): number {
  return displayDecimalsBy((threshold) => absValue >= Number(threshold));
}

export function display(value: number): string {
  const formatted = value.toFixed(displayDecimals(Math.abs(value)));
  return isFormattedZero(formatted) ? "0" : displayFormattedString(formatted);
}

export function toDecimalValue(value: bigint, decimals: number): number {
  return Number(value) / 10 ** decimals;
}

/**
 * Format a number as a percentage with exactly 2 decimal places (truncated).
 *
 * When padding is true (default), always shows 2 decimal places (e.g. "12.30").
 * When false, trailing zeros are trimmed (e.g. "12.3").
 */
export function displayPct(value: number, padding?: boolean): string {
  const pad = padding ?? true;
  const truncated = Math.trunc(value * 100) / 100;

  if (pad) {
    return displayFormattedString(truncated.toFixed(2));
  } else {
    const formatted = truncated.toFixed(2);
    const trimmed = formatted.replace(/\.?0+$/, "");
    return displayFormattedString(trimmed);
  }
}

export function fromDecimalValue(value: number, decimals: number): bigint {
  return BigInt(Math.trunc(value * 10 ** decimals));
}

File: python/src/lightcone_sdk/shared/fmt/num.py
"""Number formatting helpers mirroring Rust's shared/fmt/num.rs."""

from decimal import Decimal

from .constants import display_decimals_by


def display_formatted_string(formatted: str) -> str:
    """Add thousands separators while preserving fractional digits."""
    if not formatted:
        return "0"

    sign = ""
    if formatted.startswith("-"):
        sign = "-"
        formatted = formatted[1:]

    integer_part, dot, fraction_part = formatted.partition(".")
    try:
        integer_formatted = f"{int(integer_part):,}"
    except ValueError:
        integer_formatted = integer_part

    if dot:
        return f"{sign}{integer_formatted}.{fraction_part}"
    return f"{sign}{integer_formatted}"


def _is_formatted_zero(formatted: str) -> bool:
    value = formatted[1:] if formatted.startswith("-") else formatted
    integer_part, dot, fraction_part = value.partition(".")
    return (
        bool(integer_part)
        and all(char == "0" for char in integer_part)
        and (not dot or all(char == "0" for char in fraction_part))
    )


def display_default_formatted_string(formatted: str) -> str:
    if _is_formatted_zero(formatted):
        return "0"
    return display_formatted_string(formatted)


def _display_decimals(abs_value: float) -> int:
    return display_decimals_by(lambda threshold: abs_value >= float(threshold))


def display(amount: float) -> str:
    """Format a float for display with Rust-style decimal selection."""
    decimals = _display_decimals(abs(amount))
    return display_default_formatted_string(f"{amount:.{decimals}f}")


def display_with_decimals(amount: float, decimals: int) -> str:
    return display_formatted_string(f"{amount:.{decimals}f}")


def to_decimal_value(value: int, decimals: int) -> float:
    return value / (10 ** decimals)


def display_pct(value: float, padding: bool | None = None) -> str:
    """Format a float as a percentage with exactly 2 decimal places (truncated).

    When padding is True (default), always shows 2 decimal places (e.g. "12.30").
    When False, trailing zeros are trimmed (e.g. "12.3").
    """
    if padding is None:
        padding = True
    import math
    truncated = math.trunc(value * 100) / 100

    if padding:
        return display_formatted_string(f"{truncated:.2f}")
    else:
        formatted = f"{truncated:.2f}"
        trimmed = formatted.rstrip("0").rstrip(".")
        return display_formatted_string(trimmed)


def from_decimal_value(value: float, decimals: int) -> int:
    return int(Decimal(str(value)) * (Decimal(10) ** decimals))
